pkg/controllers/leasegarbagecollection/controller.go

```diff
@@ -0,0 +1,67 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package leasegarbagecollection
+
+import (
+	"context"
+
+	v1 "k8s.io/api/coordination/v1"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	corecontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+)
+
+// Controller for the resource
+type Controller struct {
+	kubeClient client.Client
+}
+
+// NewController is a constructor
+func NewController(kubeClient client.Client) corecontroller.Controller {
+	return corecontroller.Typed[*v1.Lease](kubeClient, &Controller{
+		kubeClient: kubeClient,
+	})
+}
+
+func (c *Controller) Name() string {
+	return "lease.garbagecollection"
+}
+
+// Reconcile the resource
+func (c *Controller) Reconcile(ctx context.Context, l *v1.Lease) (reconcile.Result, error) {
+	if l.OwnerReferences != nil {
+		return reconcile.Result{}, nil
+	}
+	err := c.kubeClient.Delete(ctx, l)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Builder(_ context.Context, m manager.Manager) corecontroller.Builder {
+	return corecontroller.Adapt(controllerruntime.
+		NewControllerManagedBy(m).
+		WithEventFilter(predicate.GenerationChangedPredicate{}).
```

Comment:
you should pre-filter out any leases that aren't the kube-node-lease so that we don't accidentally mess with any other leases in the cluster.

Reply:
This saves on performance but we should still apply these same filtering characteristics in the Reconcile as well